proto/Makefile

```diff
@@ -16,18 +16,34 @@
 # Author: Spencer Kimball (spencer.kimball@gmail.com)
 # Author: Andrew Bonventre (andybons@gmail.com)
 
+REPO_ROOT      := $(realpath ..)
+ORG_ROOT       := $(REPO_ROOT)/..
+GITHUB_ROOT    := $(ORG_ROOT)/..
+GOGOPROTO_ROOT := $(GITHUB_ROOT)/gogo/protobuf
+
+ENGINE_ROOT := $(REPO_ROOT)/storage/engine
+
 # Ensure we only have one entry in GOPATH
-GOPATH := $(shell pwd)/../../../../..
+GOPATH := $(GITHUB_ROOT)/../..
+#                                   ^  ^  ^~ GOPATH
+#                                   |  |~ GOPATH/src
+
+GOPATH_BIN := $(GOPATH)/bin
+
+PROTOC          := $(GOPATH_BIN)/protoc
+PROTOC_GEN_GOGO := $(GOPATH_BIN)/protoc-gen-gogo
+GOGO_PROTO      := $(GOGOPROTO_ROOT)/gogoproto/gogo.proto
+PROTO_PATH      := $(GOGOPROTO_ROOT):$(GOGOPROTO_ROOT)/protobuf
+ROACH_PROTOS    := $(sort $(wildcard *.proto))
+PREFIXED_PROTOS := $(ROACH_PROTOS:%=cockroach/proto/%)
+
+ROACH_CPP_PROTOS  := $(filter %api.proto %data.proto %internal.proto %config.proto %errors.proto, $(PREFIXED_PROTOS))
+ROACH_CPP_HEADERS := $(ROACH_CPP_PROTOS:cockroach/proto/%.proto=$(ENGINE_ROOT)/cockroach/proto/%.pb.h)
+ROACH_CPP_SOURCES := $(ROACH_CPP_PROTOS:cockroach/proto/%.proto=$(ENGINE_ROOT)/%.pb.cc)
 
-PROTOC      := $(GOPATH)/bin/protoc
-PROTOS      := $(sort $(wildcard *.proto))
-PROTO_GO    := $(PROTOS:.proto=.pb.go)
-GOGO_PROTOS := $(GOPATH)/src/github.com/gogo/protobuf/gogoproto/gogo.proto
-SOURCES     := $(PROTOS:.proto=.pb.cc) github.com/gogo/protobuf/gogoproto/gogo.pb.cc
-HEADERS     := $(PROTOS:.proto=.pb.h) github.com/gogo/protobuf/gogoproto/gogo.pb.h
+ROACH_GO_SOURCES  := $(ROACH_PROTOS:%.proto=%.pb.go)
 
-PROTO_PATH := ../..:$(GOPATH)/src/github.com/gogo/protobuf:$(GOPATH)/src/github.com/gogo/protobuf/protobuf
-PROTOC_GEN_GOGO := $(GOPATH)/bin/protoc-gen-gogo
+all: $(ROACH_CPP_HEADERS) $(ROACH_CPP_SOURCES) $(ROACH_GO_SOURCES)
```

Comment:
not saying it doesn't (guess we're not going to have a file named "all"), but (probably?) phony targets should get the PHONY just for phone, er, fun.
https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html